src/bip39.rs

```diff
@@ -10,152 +10,363 @@ type Passphrase = str;
     let m = m.chars().nfkd().collect::<String>();
 
     let mut salt = String::with_capacity("mnemonic".len() + p.len());
     salt.push_str("mnemonic");
     salt.push_str(p);
     let salt = salt.nfkd().collect::<String>();
 
-    crate::kdfs::pbkdf::PBKDF2_HMAC_SHA512(m.as_bytes(), salt.as_bytes(), 2048, s)
+    // unwrapping here is safe since PBKDF2_HMAC_SHA512 is only expected to fail when iteration
+    // count is zero
+    crate::kdfs::pbkdf::PBKDF2_HMAC_SHA512(m.as_bytes(), salt.as_bytes(), 2048, s).unwrap();
+}
+
+pub mod wordlist {
+    use super::*;
+    use alloc::vec::Vec;
+
+    pub type Wordlist<'a> = &'a [&'a str; 2048];
+
+    #[cfg(feature = "bip39-en")]
+    include!("bip39.en.rs");
+
+    #[cfg(feature = "bip39-jp")]
+    include!("bip39.jp.rs");
+
+    #[derive(Debug, PartialEq)]
+    pub enum Error {
+        InvalidEntropyCount(usize),
+        NoSuchWord(String),
+        ChecksumMismatch,
+    }
+
+    const fn cs(ent: usize) -> usize {
+        ent / 32
+    }
+
+    #[allow(non_snake_case)]
+    #[allow(clippy::many_single_char_names)]
+    pub fn encode(data: &[u8], wordlist: Wordlist) -> Result<String, Error> {
+        let ENT = data.len() * 8;
+
+        if ENT != 128 && ENT != 160 && ENT != 192 && ENT != 224 && ENT != 256 {
+            return Err(Error::InvalidEntropyCount(ENT));
+        }
+
+        let mut CS = [0; 32];
+        crate::hashes::sha::SHA256(data, &mut CS);
+
+        let mut ms = None;
+
+        let b = |i: usize| {
+            if i < data.len() {
+                Some(data[i] as usize)
+            } else if i - data.len() < CS.len() {
+                Some(CS[i - data.len()] as usize)
+            } else {
+                None
+            }
+        };
+
+        let mut i = 0;
+        loop {
+            if i == ENT + cs(ENT) {
+                return Ok(ms.unwrap());
+            }
+
+            let k = i / 8;
+            let r = i % 8;
+            let idx = if 16 - r > 11 {
+                match (b(k), b(k + 1)) {
+                    (Some(b0), Some(b1)) => {
+                        let x = 11 - (8 - r);
+                        let mut y = (b0 & ((1 << (8 - r)) - 1)) << x;
+                        y |= b1 >> (8 - x);
+                        y
+                    }
+                    _ => return Ok(ms.unwrap()),
+                }
+            } else {
+                match (b(k), b(k + 1), b(k + 2)) {
+                    (Some(b0), Some(b1), Some(b2)) => {
+                        let x = 11 - 8 - (8 - r);
+                        let mut y = (b0 & ((1 << (8 - r)) - 1)) << (8 + x);
+                        y |= b1 << x;
+                        y |= b2 >> (8 - x);
+                        y
+                    }
+                    _ => return Ok(ms.unwrap()),
+                }
+            };
+
+            match ms {
+                None => ms = Some(wordlist[idx].to_string()),
+                Some(ref mut ms) => {
+                    ms.push(' ');
```

Comment:
Unfortunately, Japanese is special as it uses a different [separator](https://github.com/bitcoin/bips/blob/master/bip-0039/bip-0039-wordlists.md#japanese). The test-vectors from https://github.com/bip32JP/bip32JP.github.io/blob/master/test_JP_BIP39.json contain this separator, but since in the tests only the normalized string is compared they can still pass.

I am not sure, how to best deal with this:
- Is it worth to make the word separator language depended?
- Only support English? Since localization is difficult and the seed does depend on the Language (when generated from entropy) almost all applications will only use English anyways.
- Ignore the Japanese space? (while this is technically not 100% correct)

Reply:
Ah, good point.

I kinda like the idea of extending the wordlist type with the separator, probably won't impact users since they'll most likely use the static lists that are defined in the same module.

Reply:
Can you help me figure out the Unicode code point for the jp separator? The link you gave only have the utf-8 encoding (if I read it correctly.)

Reply:
Does this help: https://www.fileformat.info/info/unicode/char/3000/index.htm

Reply:
I keep the attempted fix here (don't have the time right now to figure out why the tests aren't passing): https://github.com/iotaledger/crypto.rs/commit/5035ebdb5b483b05bd601ecf86d371b089d03bf3

Reply:
I looked a bit at the bytes from the tests with that separator, and it seems like it's only the last one or two code points that differ. This feels like someone who has dealt with Japanese encodings before will just know. (I'm way too curious so I'm currently eyeballing the trezor implementation for clues :smile: )

But I suggest we move forward like this:
1. use the new Wordlist type that has the separator per language so that there's at least a clear way forward for someone that knows how to address this
2. only check the generated mnemonics when wordlist is not Japanese in the test vectors test, and mark that skipped par with a marked TODO
3. raise an issue that can track said TODO
4. merge if you don't see other issues

I'm not really happy with removing the Japanese test vectors completely at this stage as those are the ones that exercise the normalization in the decode/verify and mnemonic_to_seed functions. Unless you have another set of test vectors for some language more complex than English but less complex than Japanese up your sleeve?

Reply:
Did a quick comparison and the Japanese wordlist from [bitcoin/bips](https://raw.githubusercontent.com/bitcoin/bips/master/bip-0039/japanese.txt) and [bip32JP](https://github.com/bip32JP/bip32JP.github.io/blob/0d1ac71933458f08a56d6ab44dd0f251158cc865/js/bip39.js#L11) are identical, so that's not an issue.

Reply:
I remember that the Japanese test vectors also gave me a very hard time for https://github.com/Wollac/iota-crypto-demo/tree/master/pkg/bip39 (and I cheated with the tests :see_no_evil: ).
@rootmos maybe this is also the issue you are having:
Voiced consonants in Japanese Hiragana are denoted by a Dakuten (゛) and in Unicode there seems to be the possibility to either encode this as a combination of two characters or as one. For example:
- https://www.fileformat.info/info/unicode/char/305e/index.htm **or**
- https://www.fileformat.info/info/unicode/char/305d/index.htm plus https://www.fileformat.info/info/unicode/char/3099/index.htm 

Changing the wordlist to use the Dakuten and similar in the matching format as the tests would fix this. But I am not sure what the recommended / normalized Unicode way is? Maybe this was lost because of different encodings of wordlist vs tests on GitHub?

Reply:
I actually tried to be very careful to only use wget, sed, grep to create the wordlist files, not even open them in any kind of editor. But I agree that something like that is probably the cause.

I've pushed the "half-way" solution now: https://github.com/iotaledger/crypto.rs/pull/31/commits/5670f988e03a959b9d6942c3bbbd08c3b8cb3619 together with a warning in its documentation.

Can you do a final review and perhaps create the issue referencing this "hole" in the current implementation?

Reply:
This is definitely a "stinker area", I also suggest we ask someone else to do another clean room implementation and compare/replace the one I've written here.

Reply:
I did found this commit https://github.com/bip32JP/bip32JP.github.io/commit/360c05a6439e5c461bbe5e84c7567ec38eb4ac5f saying `test vectors should be NOT normalized. Too many people relying on misconception that user INPUT will be == normalized text. If your library fails on this input, it will fail in real world.`
Previous to this commit the mnemonics have all been normalized and the tests pass.
Having non-normalized passphrase and mnemonic makes a lot of sense when you want to test the mnemonic->seed functionality. However, I guess for entropy->mnemonics I guess the best we can do is normalize before comparing.

Reply:
Which is what the code is doing now.